lib/ProductOpener/Ingredients.pm

```diff
@@ -863,13 +865,146 @@ my %ignore_strings_after_percent = (
 );
 
 
+=head2 has_specific_ingredient_property ( product_ref, ingredient_id, property )
+
+Check if the specific ingredients structure (extracted from the end of the ingredients list and product labels)
+contains a property for an ingredient. (e.g. do we have an origin specified for a specific ingredient)
+
+If the ingredient_id parameter is undef, then we return the value for any specific ingredient.
+(useful for products for which we do not have ingredients, but for which we have a label like "French eggs":
+we can still derive the origin of the ingredients, e.g. for the Eco-Score)
+
+=head3 Return values
+
+=head4 value
+
+- undef if we don't have a specific ingredient with the requested property matching the requested ingredient
+- otherwise the value for the matching specific ingredient
+
+=cut
+
+sub has_specific_ingredient_property($$$) {
+
+	my $product_ref = shift;
+	my $ingredient_id = shift;
+	my $property = shift;
+
+	my $value;
+
+	# If we have specific ingredients, check if we have a higher fruits / vegetables content
+	if (defined $product_ref->{specific_ingredients}) {
+		foreach my $ingredient_ref (@{$product_ref->{specific_ingredients}}) {
+			my $specific_ingredient_id = $ingredient_ref->{id};
+			if ((defined $ingredient_ref->{origins})
+				and ((not defined $ingredient_id) or (is_a("ingredients", $ingredient_id, $specific_ingredient_id)))) {
+
+				$value = $ingredient_ref->{origins};
+				last;
```

Comment:
I'm asking myself in if this kind of treatment we should not continue for sanity checks. Something like:

```perl
defined $value and $value = $ingredient_ref->{origins} 
    or log->warning("Got more than one specific ingredient origin ", {product => $product_ref->{id}, ingredient_id => $ingredient_id});
```

Reply:
good idea. I also just noticed that I hardcoded "origins" in this function even though it's passed a $property parameter.